files/zh-cn/web/api/web_components/using_custom_elements/index.md

```diff
@@ -1,264 +1,529 @@
 
-共有两种 custom elements：
+`define()` 方法接受以下参数：
 
-- **Autonomous custom elements** 是独立的元素，它不继承其他内建的 HTML 元素。你可以直接把它们写成 HTML 标签的形式，来在页面上使用。例如 `<popup-info>`，或者是`document.createElement("popup-info")`这样。
-- **Customized built-in elements** 继承自基本的 HTML 元素。在创建时，你必须指定所需扩展的元素（正如上面例子所示），使用时，需要先写出基本的元素标签，并通过 [`is`](/zh-CN/docs/Web/HTML/Global_attributes#is) 属性指定 custom element 的名称。例如`<p is="word-count">`, 或者 `document.createElement("p", { is: "word-count" })`。
+- `name`
+  - : 元素的名称。必须以小写字母开头，包含一个连字符，并符合规范中有效名称的定义中列出的一些其他规则。
+- `constructor`
+  - : 自定义元素的构造函数。
+- `options`
+  - : 仅对于自定义内置元素，这是一个包含单个属性 `extends` 的对象，该属性是一个字符串，命名了要扩展的内置元素。
 
-## 示例
+例如，以下代码注册了名为 `WordCount` 的自定义内置元素：
+
+```js
+customElements.define("word-count", WordCount, { extends: "p" });
+```
+
+以下代码注册了名为 `PopupInfo` 的独立自定义元素：
+
+```js
+customElements.define("popup-info", PopupInfo);
+```
+
+## 使用自定义元素
+
+一旦你定义并注册了自定义元素，就可以在代码中使用它。
+
+要使用自定义内置元素，请使用内置元素，但将自定义名称作为 [`is`](/zh-CN/docs/Web/HTML/Global_attributes/is) 属性的值：
+
+```html
+<p is="word-count"></p>
+```
+
+要使用独立自定义元素，就像使用内置的 HTML 元素一样，使用自定义名称即可：
+
+```html
+<popup-info>
+  <!-- 元素的内容 -->
+</popup-info>
+```
+
+## 响应属性变化
+
+与内置元素一样，自定义元素可以使用 HTML 属性来配置元素的行为。为了有效地使用属性，元素必须能够响应属性值的变化。为此，自定义元素需要将以下成员添加到实现自定义元素的类中：
 
-让我们来看几个简单示例，来了解如何创建 custom elements。
+- 一个名为 `observedAttributes` 的静态属性。这必须是一个包含元素需要变更通知的所有属性名称的数组。
+- `attributeChangedCallback()` 生命周期回调的实现。
 
-### Autonomous custom elements
+`attributeChangedCallback()` 回调在列在元素的 `observedAttributes` 属性中的属性被添加、修改、移除或替换时调用。
 
-我们来看一下 [`<popup-info-box>`](https://github.com/mdn/web-components-examples/tree/master/popup-info-box-web-component) (查看[在线示例](https://mdn.github.io/web-components-examples/popup-info-box-web-component/))，一个关于 autonomous custom element 的例子。它包含有一个图标和一段文字，并且图标显示在页面上。在这个图标获取焦点时，它会显示一个包含该段文字的信息框，用于展示更多的信息。
+回调接受三个参数：
 
-为了实现这个功能，首先创建一个 JavaScript 文件，定义一个叫做`PopUpInfo`的类，它继承自{{domxref("HTMLElement")}}。Autonomous custom elements 总是继承自`HTMLElement`。
+- 发生变化的属性的名称。
+- 属性的旧值。
+- 属性的新值。
+
+例如，下面这个独立自定义元素将观察一个 `size` 属性，并在它们发生变化时记录旧值和新值：
 
 ```js
-class PopUpInfo extends HTMLElement {
+// 为这个元素创建类
+class MyCustomElement extends HTMLElement {
+  static observedAttributes = ["size"];
+
   constructor() {
-    // 必须首先调用 super 方法
     super();
+  }
 
-    // 元素的功能代码写在这里
-
-    ...
+  attributeChangedCallback(name, oldValue, newValue) {
+    console.log(
+      `Attribute ${name} has changed from ${oldValue} to ${newValue}.`,
+    );
   }
 }
+
+customElements.define("my-custom-element", MyCustomElement);
 ```
 
-上述代码片段中，类的构造函数{{jsxref("Classes.constructor","constructor")}}总是先调用[`super()`](/zh-CN/docs/Web/JavaScript/Reference/Operators/super)来建立正确的原型链继承关系。
+请注意，如果元素的 `HTML` 声明包含一个被观察的属性，那么在元素的声明首次解析时，`attributeChangedCallback()` 将被调用，即在属性被初始化后。因此，在以下示例中，即使属性再也没有被更改，当 DOM 被解析时，`attributeChangedCallback()` 也会被调用：（译注：即每个直接在 HTML 中声明的且被观察的属性会立即触发 `attributeChangedCallback()` 一次）
```

Comment:
Let's remove all the translation note, if the body is not will written, please update the original English content.

```suggestion
请注意，如果元素的 `HTML` 声明包含一个被观察的属性，那么在元素的声明首次解析时，`attributeChangedCallback()` 将被调用，即在属性被初始化后。因此，在以下示例中，即使属性再也没有被更改，当 DOM 被解析时，`attributeChangedCallback()` 也会被调用：
```